### PR TITLE
Fetch operation

### PR DESCRIPTION
**Pull request by seut**

### CODE REVIEW

File: sql/src/main/java/io/crate/executor/transport/TransportFetchNodeAction.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+                request.toFetchReferences(),
+                request.closeContext(),
+                collectContextService,
+                threadPool,
+                functions,
+                ramAccountingContext);
+
+        ListenableFuture<Bucket> result;
+        try {
+            result = fetchOperation.fetch();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/TransportFetchNodeAction.java]: imho a bit strange that .fetch can fail itself and then the future could also fail too

File: sql/src/main/java/io/crate/operation/merge/MergeOperation.java [hunk 2]
```diff
@@ -21,83 +21,78 @@
...
 
-import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.Executor;
 
 /**
  * merge rows - that's it
  */
-public class MergeOperation implements DownstreamOperation {
+@Singleton
+public class MergeOperation {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/merge/MergeOperation.java]: please rename to PageDownstreamFactory or something like that.. this isn't a MergeOperation anymore.

File: sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java [hunk 3]
```diff
@@ -60,6 +49,8 @@
  */
 public class LuceneDocCollector extends Collector implements CrateCollector, RowUpstream {
 
+    public final static int PAGE_SIZE = 10000;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java]: isn't there a better place for the PAGE_SIZE? It shouldn't be Lucene specific, right?

File: sql/src/main/java/io/crate/operation/fetch/FetchAbortedException.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.operation.fetch;
+
+import java.io.IOException;
+
+public class FetchAbortedException extends IOException {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/fetch/FetchAbortedException.java]: shouldn't this have some kind of default message ?

File: sql/src/main/java/io/crate/operation/fetch/LuceneDocFetcher.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                .searchContext(searchContext)
+                .searchLookup(searchContext.lookup(false));
+        for (LuceneCollectorExpression<?> collectorExpression : collectorExpressions) {
+            collectorExpression.startCollect(collectorContext);
+        }
+        visitorEnabled = fieldsVisitor.required();
+
+        try {
+            for (int index = 0; index < shardDocIdsBucket.size(); index++) {
+                int docId = shardDocIdsBucket.docId(index);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/fetch/LuceneDocFetcher.java]: isn't this the same as in the LuceneDocCollector? Is it possible to share something here?

File: sql/src/main/java/io/crate/operation/projectors/NoOpProjector.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import io.crate.operation.RowUpstream;
+
+/**
+ * a projector that does nothing.
+ * Can be used by classes that implement ProjectorUpstream to have a "default" downstream
+ * to avoid null checks
+ */
+public class NoOpProjector implements Projector, RowDownstreamHandle {
+
+    public static NoOpProjector INSTANCE = new NoOpProjector();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/projectors/NoOpProjector.java]: should be final

File: sql/src/main/java/io/crate/operation/fetch/FetchAbortedException.java (hunk 4 above)
Comment (seut) [inline sql/src/main/java/io/crate/operation/fetch/FetchAbortedException.java]: why? this will never bubble...

**Comment by mfussenegger** [comment]
ack

**Pull request merged**